### PR TITLE
MultiRow Insert Optimization

### PR DESCRIPTION
**Pull request by ashudeep-sharma**
This Pull request contains the changes for Optimising the vindex based queries on look-up shard for Multi-Row Insert.
@sougou , Please take a look.

### CODE REVIEW

File: go/vt/vtgate/router.go [hunk 1]
```diff
@@ -701,57 +704,86 @@ func (rtr *Router) handleGenerate(vcursor *queryExecutor, gen *engine.Generate)
 	return insertid, nil
 }
 
-func (rtr *Router) handlePrimary(vcursor *queryExecutor, vindexKey interface{}, colVindex *vindexes.ColumnVindex, bv map[string]interface{}, rowNum int) (ksid []byte, err error) {
-	if vindexKey == nil {
+func (rtr *Router) handlePrimary(vcursor *queryExecutor, vindexKeys []interface{}, colVindex *vindexes.ColumnVindex, bv map[string]interface{}, allShards []*topodatapb.ShardReference) (keyspaceIDs [][]byte, err error) {
+	if vindexKeys == nil {
```
Comment (sougou) [inline go/vt/vtgate/router.go]: I don't think this is possible. Were you able to get coverage to trigger this path?

File: go/vt/vtgate/router.go [hunk 2]
```diff
@@ -701,57 +704,86 @@ func (rtr *Router) handleGenerate(vcursor *queryExecutor, gen *engine.Generate)
...
-	bv["_"+colVindex.Column.CompliantName()+strconv.Itoa(rowNum)] = vindexKey
-	return ksid, nil
+	return keyspaceIDs, nil
 }
 
-func (rtr *Router) handleNonPrimary(vcursor *queryExecutor, vindexKey interface{}, colVindex *vindexes.ColumnVindex, bv map[string]interface{}, ksid []byte, rowNum int) error {
+func (rtr *Router) handleNonPrimary(vcursor *queryExecutor, vindexKeys []interface{}, colVindex *vindexes.ColumnVindex, bv map[string]interface{}, ksids [][]byte) error {
 	if colVindex.Owned {
-		if vindexKey == nil {
+		if vindexKeys == nil {
```
Comment (sougou) [inline go/vt/vtgate/router.go]: This should also not be possible.

**Comment by sougou** [comment]
Nice work. Only couple of comments.

**Comment by ashudeep-sharma** [comment]
Removed the redundant code in the latest commit.
Also after running code coverage analysis for vindexes, vtgate and planbuilder, i got the following coverage stats:

coverage: **82.8%** of statements
ok  	_github.com/youtube/vitess/go/vt/vtgate/vindexes_	0.024s

coverage: **80.3%** of statements
ok  	_github.com/youtube/vitess/go/vt/vtgate_	0.075s

coverage: **98.4%** of statements
ok  	_github.com/youtube/vitess/go/vt/vtgate/planbuilder_	0.068s

File: go/vt/vtgate/vindexes/lookup_hash.go [hunk 3]
```diff
@@ -97,14 +97,14 @@ func (vind *LookupHashUnique) Map(vcursor VCursor, ids []interface{}) ([][]byte,
 	return vind.lkp.MapUniqueLookup(vcursor, ids)
 }
 
-// Verify returns true if id maps to ksid.
-func (vind *LookupHashUnique) Verify(vcursor VCursor, id interface{}, ksid []byte) (bool, error) {
-	return vind.lkp.Verify(vcursor, id, ksid)
+// Verify returns true if ids maps to ksids.
+func (vind *LookupHashUnique) Verify(vcursor VCursor, ids []interface{}, ksids [][]byte) (bool, error) {
+	return vind.lkp.Verify(vcursor, ids, ksids)
```
Comment (sougou) [inline go/vt/vtgate/vindexes/lookup_hash.go]: Needs test case. Same with the rest of the functions.

File: go/vt/vtgate/vindexes/lookup_internal.go [hunk 4]
```diff
@@ -13,18 +16,18 @@ type lookup struct {
 	isHashedIndex      bool
 }
 
-func (lkp *lookup) Init(m map[string]string, isHashed bool) {
-	t := m["table"]
-	from := m["from"]
-	to := m["to"]
+func (lkp *lookup) Init(lookupQueryParams map[string]string, isHashed bool) {
```
Comment (sougou) [inline go/vt/vtgate/vindexes/lookup_internal.go]: Many functions in this file are missing critical code coverage. It's probably an artifact of previous changes.

File: go/vt/vtgate/engine/route.go [hunk 5]
```diff
@@ -744,57 +747,80 @@ func (route *Route) handleGenerate(vcursor VCursor, queryConstruct *queryinfo.Qu
...
 				return fmt.Errorf("value must be supplied for column %v", colVindex.Column)
 			}
-			var err error
-			vindexKey, err = reversible.ReverseMap(vcursor, ksid)
+			vindexKeys, err = reversible.ReverseMap(vcursor, reverseKsids)
 			if err != nil {
 				return err
 			}
-			if vindexKey == nil {
+			if vindexKeys == nil {
```
Comment (sougou) [inline go/vt/vtgate/engine/route.go]: Is this possible?

File: go/vt/vtgate/engine/route.go [hunk 6]
```diff
@@ -744,57 +747,80 @@ func (route *Route) handleGenerate(vcursor VCursor, queryConstruct *queryinfo.Qu
...
+		if reverseKsids != nil {
 			reversible, ok := colVindex.Vindex.(vindexes.Reversible)
 			if !ok {
 				return fmt.Errorf("value must be supplied for column %v", colVindex.Column)
 			}
-			var err error
-			vindexKey, err = reversible.ReverseMap(vcursor, ksid)
+			vindexKeys, err = reversible.ReverseMap(vcursor, reverseKsids)
 			if err != nil {
 				return err
```
Comment (sougou) [inline go/vt/vtgate/engine/route.go]: Missing test case for this. Is this possible?

File: go/vt/vtgate/vindexes/binary.go [hunk 7]
```diff
@@ -25,13 +25,21 @@ func (vind *Binary) Cost() int {
...
+		return false, fmt.Errorf("Binary.Verify: length of ids %v doesn't match length of ksids %v", len(ids), len(ksids))
 	}
-	return bytes.Compare(data, ksid) == 0, nil
+	for rowNum := range ids {
+		data, err := getBytes(ids[rowNum])
+		if err != nil {
+			return false, fmt.Errorf("Binary.Verify: %v", err)
+		}
+		if bytes.Compare(data, ksids[rowNum]) != 0 {
+			return false, nil
```
Comment (ashudeep-sharma) [inline go/vt/vtgate/vindexes/binary.go]: Done

File: go/vt/vtgate/vindexes/lookup_hash.go (hunk 3 above)
Comment (ashudeep-sharma) [inline go/vt/vtgate/vindexes/lookup_hash.go]: Done

**Comment by ashudeep-sharma** [comment]
@sougou : Added corresponding positive as well as negative test-cases for all the vindexes. Please take a look

**Comment by sougou** [comment]
LGTM

**Pull request merged**